In [1]:
import pandas as pd
import geopandas as gpd
import matplotlib.pyplot as plt
plt.rcParams['font.sans-serif'] = ['Microsoft YaHei', 'SimHei']  # 优先微软雅黑/黑体
plt.rcParams['axes.unicode_minus'] = False  # 修复负号显示
data=pd.read_csv(r"D:\MachineLearningTrain\zf\Export_Output3.txt",sep=',',index_col=0,encoding='utf-8-sig')
data

CID  正样本     POINT_X    POINT_Y  灾害体          灾害_1  \
OBJECTID                                                             
1         2292107.0    0  116.161667  27.052222  NaN           NaN   
2         2286798.0    0  116.162500  27.060556  NaN           NaN   
3         2263128.0    0  116.227222  27.084722  NaN           NaN   
4         2262623.0    0  116.173056  27.085000  NaN           NaN   
5         2246373.0    0  116.167222  27.098056  NaN           NaN   
...             ...  ...         ...        ...  ...           ...   
421             NaN    1  116.347500  27.660556   滑坡  3.610260e+15   
422             NaN    1  116.350000  27.668333   滑坡  3.610260e+15   
423             NaN    1  116.288056  27.631389   滑坡  3.610260e+15   
424             NaN    1  116.317222  27.623889   滑坡  3.610260e+15   
425             NaN    1  116.405556  27.630000   滑坡  3.610260e+15   

                 灾害_12   灾害__13      野外编         室内编  ... EucDist_shp9  \
OBJECTID                                              ...                
1                  NaN      NaN      NaN         NaN  ...          NaN   
2                  NaN      NaN      NaN         NaN  ...          NaN   
3                  NaN      NaN      NaN         NaN  ...     0.012744   
4                  NaN      NaN      NaN         NaN  ...          NaN   
5                  NaN      NaN      NaN         NaN  ...          NaN   
...                ...      ...      ...         ...  ...          ...   
421       3.610260e+15  刘光荣房后滑坡  YH-1142  344402-025  ...          NaN   
422       3.610260e+15  丁友华房后滑坡  YH-2165  344402-026  ...          NaN   
423       3.610260e+15     磁坑滑坡  YH-1141  344401-048  ...     0.007514   
424       3.610260e+15  纪焕盛屋侧滑坡  YH-1137  344401-070  ...          NaN   
425       3.610260e+15  张荣辉房后滑坡  YH-1152  344402-113  ...     0.028853   

              ndvi 宜黄dem            曲率          坡向         坡度  \
OBJECTID                                                        
1         0.790292   553 -9.072000e+09  304.939301  19.850380   
2         0.791342   501  6.480000e+09  275.527527   7.196245   
3         0.770125   431 -1.814400e+10   26.200115  15.889066   
4         0.731075   333 -3.888000e+09  195.255127   2.646518   
5         0.752542   439  1.166400e+10   45.763897  23.269426   
...            ...   ...           ...         ...        ...   
421       0.630142    97 -1.036800e+10  291.297363  14.398109   
422       0.723758   115  9.072000e+09   72.149681  14.105568   
423       0.762558   133 -1.684800e+10   57.528809  17.596445   
424       0.788492   177 -0.000000e+00  170.362457  12.294831   
425       0.749033   120  1.425600e+10  218.829819  12.044478   

          yihuangjiangshui  NEAR_FID  Feature_yanxing  Feature_dimao  
OBJECTID                                                              
1              1781.722900         9         0.303340       0.303340  
2              1768.051392         9         0.303340       0.303340  
3              1778.674316         9         0.303340       0.303340  
4              1737.857178         9         0.303340       0.303340  
5              1786.058594        28         0.000949       0.000949  
...                    ...       ...              ...            ...  
421            1676.948486         7         0.002310       0.002310  
422            1676.948486         2         0.006038       0.006038  
423            1690.098511         8         0.018984       0.018984  
424            1705.154297         8         0.018984       0.018984  
425            1701.982910         8         0.018984       0.018984  

[425 rows x 48 columns]

In [2]:
Y=data.iloc[:,data.columns =="正样本"]
Y

正样本
OBJECTID     
1           0
2           0
3           0
4           0
5           0
...       ...
421         1
422         1
423         1
424         1
425         1

[425 rows x 1 columns]

删除无关特征

In [3]:
columns_to_drop = data.columns[4:37]
columns_to_drop

Index(['灾害体', '灾害_1', '灾害_12', '灾害__13', '野外编', '室内编', '省', '市', '县', '乡镇',
       '地理位', '经度', '纬度', 'X', 'Y', 'Z', '坡顶最', '坡底最', '治理情', '地下水', '构造部',
       '地震烈', '目前稳', '今后变', '威胁人', '威胁财', '灾害等', '险情等', '防治建', '隐患点', '监测建',
       '毁坏房', '毁路'],
      dtype='object')

In [4]:
data2 = data.drop(columns=columns_to_drop)
data2

CID  正样本     POINT_X    POINT_Y  EucDist_shp10  EucDist_shp9  \
OBJECTID                                                                       
1         2292107.0    0  116.161667  27.052222       0.260116           NaN   
2         2286798.0    0  116.162500  27.060556            NaN           NaN   
3         2263128.0    0  116.227222  27.084722            NaN      0.012744   
4         2262623.0    0  116.173056  27.085000            NaN           NaN   
5         2246373.0    0  116.167222  27.098056            NaN           NaN   
...             ...  ...         ...        ...            ...           ...   
421             NaN    1  116.347500  27.660556            NaN           NaN   
422             NaN    1  116.350000  27.668333            NaN           NaN   
423             NaN    1  116.288056  27.631389            NaN      0.007514   
424             NaN    1  116.317222  27.623889            NaN           NaN   
425             NaN    1  116.405556  27.630000            NaN      0.028853   

              ndvi  宜黄dem            曲率          坡向         坡度  \
OBJECTID                                                         
1         0.790292    553 -9.072000e+09  304.939301  19.850380   
2         0.791342    501  6.480000e+09  275.527527   7.196245   
3         0.770125    431 -1.814400e+10   26.200115  15.889066   
4         0.731075    333 -3.888000e+09  195.255127   2.646518   
5         0.752542    439  1.166400e+10   45.763897  23.269426   
...            ...    ...           ...         ...        ...   
421       0.630142     97 -1.036800e+10  291.297363  14.398109   
422       0.723758    115  9.072000e+09   72.149681  14.105568   
423       0.762558    133 -1.684800e+10   57.528809  17.596445   
424       0.788492    177 -0.000000e+00  170.362457  12.294831   
425       0.749033    120  1.425600e+10  218.829819  12.044478   

          yihuangjiangshui  NEAR_FID  Feature_yanxing  Feature_dimao  
OBJECTID                                                              
1              1781.722900         9         0.303340       0.303340  
2              1768.051392         9         0.303340       0.303340  
3              1778.674316         9         0.303340       0.303340  
4              1737.857178         9         0.303340       0.303340  
5              1786.058594        28         0.000949       0.000949  
...                    ...       ...              ...            ...  
421            1676.948486         7         0.002310       0.002310  
422            1676.948486         2         0.006038       0.006038  
423            1690.098511         8         0.018984       0.018984  
424            1705.154297         8         0.018984       0.018984  
425            1701.982910         8         0.018984       0.018984  

[425 rows x 15 columns]

读取矢量要素（水系、道路、岩性）

In [5]:
import geopandas as gpd
import geopandas as gpd

# 读取点数据（正负样本点）
points_gdf = gpd.read_file(r"D:\MachineLearningTrain\zf\Export_Output3.shp",encoding='UTF-8')  # 替换为你的点数据路径

# 读取矢量要素（如水系、道路、岩性）
shuixi = gpd.read_file(r"D:\MachineLearningTrain\yihuangshuixi\shuixi.shp",encoding='UTF-8')
daolu = gpd.read_file(r"D:\MachineLearningTrain\yihuangdaolu\daolu.shp",encoding='UTF-8')
dimao = gpd.read_file(r"D:\MachineLearningTrain\yihuangdimao\dimao.shp",encoding='UTF-8')
yanxing = gpd.read_file(r"D:\MachineLearningTrain\yihuangyanxing\yanxing.shp",encoding='UTF-8')
points_gdf
columns_to_drop = points_gdf.columns[1:49]
columns_to_drop

Index(['CID', '正样本', 'POINT_X', 'POINT_Y', '灾害体', '灾害_1', '灾害_12', '灾害__13',
       '野外编', '室内编', '省', '市', '县', '乡镇', '地理位', '经度', '纬度', 'X', 'Y', 'Z',
       '坡顶最', '坡底最', '治理情', '地下水', '构造部', '地震烈', '目前稳', '今后变', '威胁人', '威胁财',
       '灾害等', '险情等', '防治建', '隐患点', '监测建', '毁坏房', '毁路', 'EucDist_sh',
       'EucDist__1', 'ndvi', '宜黄dem', '曲率', '坡向', '坡度', 'yihuangjia',
       'NEAR_FID', 'Feature_ya', 'Feature_di'],
      dtype='object')

In [6]:
points_gdf.drop(columns=['CID','灾害体', '灾害_1', '灾害_12', '灾害__13',
       '野外编', '室内编', '省', '市', '县', '乡镇', '地理位', '经度', '纬度', 'X', 'Y', 'Z',
       '坡顶最', '坡底最', '治理情', '地下水', '构造部', '地震烈', '目前稳', '今后变', '威胁人', '威胁财',
       '灾害等', '险情等', '防治建', '隐患点', '监测建', '毁坏房', '毁路', 'EucDist_sh',
       'EucDist__1','Feature_ya', 'Feature_di'
    ], inplace=True,errors='ignore')

In [7]:
points_gdf.head()

OBJECTID  正样本     POINT_X    POINT_Y      ndvi  宜黄dem            曲率  \
0         1    0  116.161667  27.052222  0.790292    553 -9.072000e+09   
1         2    0  116.162500  27.060556  0.791342    501  6.480000e+09   
2         3    0  116.227222  27.084722  0.770125    431 -1.814400e+10   
3         4    0  116.173056  27.085000  0.731075    333 -3.888000e+09   
4         5    0  116.167222  27.098056  0.752542    439  1.166400e+10   

         坡向        坡度  yihuangjia  NEAR_FID                    geometry  
0  304.9390  19.85040     1781.72         9  POINT (116.16167 27.05222)  
1  275.5280   7.19625     1768.05         9   POINT (116.1625 27.06056)  
2   26.2001  15.88910     1778.67         9  POINT (116.22722 27.08472)  
3  195.2550   2.64652     1737.86         9    POINT (116.17306 27.085)  
4   45.7639  23.26940     1786.06        28  POINT (116.16722 27.09806)

#空间分析:  从矢量数据提取特征    
1.计算点到水系，道路最小距离

In [8]:
from shapely.ops import nearest_points
from shapely.wkt import loads
# 计算点到最近水系的距离
def get_distance(point, target_gdf):
    distances = []
    for geom in target_gdf.geometry:
        distances.append(point.distance(geom))
    return min(distances) if distances else None

# 为每个点计算到水系的距离
points_gdf['distance_to_water'] = points_gdf.geometry.apply(
    lambda x: get_distance(x, shuixi)
)

# 同理计算到道路的距离
points_gdf['distance_to_road'] = points_gdf.geometry.apply(
    lambda x: get_distance(x, daolu)
)

In [9]:
# 空间连接：判断点是否落在岩性多边形内
rock_joined = gpd.sjoin(points_gdf, yanxing, how="left", predicate='intersects')

# 打印岩性数据（yanxing）的列名，确认实际字段名
print("岩性数据的列名：", yanxing.columns.tolist())

岩性数据的列名： ['AREA', 'PERIMETER', 'CHINA_PC_', 'CHINA_PC_I', 'SUBCLASS', 'SUBCLASS_', 'RINGS_OK', 'RINGS_NOK', 'NEWSUID', 'DOMSOIL', 'LANDFORM', 'SLOPE_MED', 'RELIEF_MED', 'ELEV_MED', 'SMU_MOD', 'PRID', 'LITHOLOGY', 'SQKM', 'INDEX', 'ISOCSUID', 'geometry']


In [10]:
# 确认岩性字段为 'LITHOLOGY'
print("岩性字段示例：", yanxing['LITHOLOGY'].unique()[:5])  # 查看前5个唯一岩性类型

岩性字段示例： ['SC4' 'SC2' 'IA1' 'SO1' 'IP4']


In [11]:
# 确保点数据（points_gdf）和岩性数据的坐标系一致
if points_gdf.crs != yanxing.crs:
    yanxing = yanxing.to_crs(points_gdf.crs)

# 空间连接：判断点是否落在岩性多边形内（sjoin空间连接）
rock_joined = gpd.sjoin(
    points_gdf,
    yanxing[['LITHOLOGY', 'geometry']],  # 仅保留岩性字段和几何
    how="left",
    predicate='intersects'  # 判断点是否与岩性多边形相交
)

In [12]:
# 提取岩性类型（此处岩性数据有字段'LITHOLOGY'）
points_gdf['rock_type'] = rock_joined['LITHOLOGY'].values

# 处理缺失值（如点不在任何岩性多边形内）
points_gdf['rock_type'] = points_gdf['rock_type'].fillna('UNKNOWN')

In [13]:
# （可选）查看岩性分布统计
print("岩性类型统计：")
print(points_gdf['rock_type'].value_counts())

岩性类型统计：
rock_type
IA1    195
SC4     95
SO1     75
SC2     43
IP4     16
UR1      1
Name: count, dtype: int64


处理岩性类别不平衡问题（IA1 195个，UR1只有1个）这里只有UR1为低频类别（大于10为高频/以10为阈值划分）

In [14]:
# 假设 UR1 对应特殊性土（如泥炭土）
#points_gdf['rock_type'] = points_gdf['rock_type'].replace({
#    'UR1': 'SPECIAL_SOIL'  # 合并为新类别 "SPECIAL_SOIL"
#})

# 或者直接删除 UR1 的样本（如果它不可靠）
points_gdf = points_gdf[points_gdf['rock_type'] != 'UR1']

In [15]:
# 重新统计
print(points_gdf['rock_type'].value_counts())

# 特征编码（独热编码）
rock_encoded = pd.get_dummies(points_gdf['rock_type'], prefix='rock')
points_gdf = pd.concat([points_gdf, rock_encoded], axis=1)

# 后续分析（如滑坡风险建模）
# ...

rock_type
IA1    195
SC4     95
SO1     75
SC2     43
IP4     16
Name: count, dtype: int64


In [16]:
points_gdf.head()

OBJECTID  正样本     POINT_X    POINT_Y      ndvi  宜黄dem            曲率  \
0         1    0  116.161667  27.052222  0.790292    553 -9.072000e+09   
1         2    0  116.162500  27.060556  0.791342    501  6.480000e+09   
2         3    0  116.227222  27.084722  0.770125    431 -1.814400e+10   
3         4    0  116.173056  27.085000  0.731075    333 -3.888000e+09   
4         5    0  116.167222  27.098056  0.752542    439  1.166400e+10   

         坡向        坡度  yihuangjia  NEAR_FID                    geometry  \
0  304.9390  19.85040     1781.72         9  POINT (116.16167 27.05222)   
1  275.5280   7.19625     1768.05         9   POINT (116.1625 27.06056)   
2   26.2001  15.88910     1778.67         9  POINT (116.22722 27.08472)   
3  195.2550   2.64652     1737.86         9    POINT (116.17306 27.085)   
4   45.7639  23.26940     1786.06        28  POINT (116.16722 27.09806)   

   distance_to_water  distance_to_road rock_type  rock_IA1  rock_IP4  \
0           0.258510          0.061480       IA1      True     False   
1           0.250265          0.053936       IA1      True     False   
2           0.240671          0.013757       IA1      True     False   
3           0.227079          0.028091       IA1      True     False   
4           0.213397          0.024769       SC4     False     False   

   rock_SC2  rock_SC4  rock_SO1  
0     False     False     False  
1     False     False     False  
2     False     False     False  
3     False     False     False  
4     False      True     False

#需删除rock_type列，只保留独热编码后的rock_IA1...（避免冗余）

In [17]:
points_gdf = points_gdf.drop(columns=['rock_type'])
# 将布尔值转换为整数（0/1）
boolean_columns = ['rock_IA1', 'rock_IP4', 'rock_SC2', 'rock_SC4', 'rock_SO1']
points_gdf[boolean_columns] = points_gdf[boolean_columns].astype(int)
points_gdf.head()

OBJECTID  正样本     POINT_X    POINT_Y      ndvi  宜黄dem            曲率  \
0         1    0  116.161667  27.052222  0.790292    553 -9.072000e+09   
1         2    0  116.162500  27.060556  0.791342    501  6.480000e+09   
2         3    0  116.227222  27.084722  0.770125    431 -1.814400e+10   
3         4    0  116.173056  27.085000  0.731075    333 -3.888000e+09   
4         5    0  116.167222  27.098056  0.752542    439  1.166400e+10   

         坡向        坡度  yihuangjia  NEAR_FID                    geometry  \
0  304.9390  19.85040     1781.72         9  POINT (116.16167 27.05222)   
1  275.5280   7.19625     1768.05         9   POINT (116.1625 27.06056)   
2   26.2001  15.88910     1778.67         9  POINT (116.22722 27.08472)   
3  195.2550   2.64652     1737.86         9    POINT (116.17306 27.085)   
4   45.7639  23.26940     1786.06        28  POINT (116.16722 27.09806)   

   distance_to_water  distance_to_road  rock_IA1  rock_IP4  rock_SC2  \
0           0.258510          0.061480         1         0         0   
1           0.250265          0.053936         1         0         0   
2           0.240671          0.013757         1         0         0   
3           0.227079          0.028091         1         0         0   
4           0.213397          0.024769         0         0         0   

   rock_SC4  rock_SO1  
0         0         0  
1         0         0  
2         0         0  
3         0         0  
4         1         0

dimao数据处理

In [18]:
import geopandas as gpd
import pandas as pd
# 查看数据结构和属性字段（确保包含地貌类型字段，如'TYPE'或'LAND_TYPE'）
print("dimao数据列：", dimao.columns)
print("前几行数据示例：")
print(dimao.head())

dimao数据列： Index(['AREA', 'PERIMETER', 'CHINA_PC_', 'CHINA_PC_I', 'SUBCLASS', 'SUBCLASS_',
       'RINGS_OK', 'RINGS_NOK', 'NEWSUID', 'DOMSOIL', 'LANDFORM', 'SLOPE_MED',
       'RELIEF_MED', 'ELEV_MED', 'SMU_MOD', 'PRID', 'LITHOLOGY', 'SQKM',
       'INDEX', 'ISOCSUID', 'geometry'],
      dtype='object')
前几行数据示例：
       AREA  PERIMETER  CHINA_PC_  CHINA_PC_I SUBCLASS  SUBCLASS_  RINGS_OK  \
0  0.006940   0.841348      53258       53331     POLY      12810         1   
1  0.015169   0.845052      53811       53886     POLY       4502         1   
2  0.006038   0.645249      53882       53957     POLY       3153         1   
3  0.033164   1.454536      53962       54037     POLY       5008         1   
4  0.002104   0.315530      53973       54048     POLY      12800         1   

   RINGS_NOK  NEWSUID DOMSOIL  ... SLOPE_MED  RELIEF_MED  ELEV_MED  SMU_MOD  \
0          0    774.0     ATc  ...       4.4        32.0      87.0   1604.0   
1          0    160.0     ACh  ...      28.9       25

此处发现地貌数据中也存在LITHOLOGY（岩性）之前提过，这里选择合并或删除一个(避免冗余)     
SLOPE_MED（坡度中位数）RELIEF_MED（地形起伏度）ELEV_MED（高程中位数）#与之前不同此处都为区域性的均值，与之前的不构成重复

AREA面积 PERIMETER周长 CHINA_PC_  CHINA_PC_I行政区划分码 SUBCLASS  SUBCLASS_地貌子类别分类编码        
DOMSOIL土壤类型 LANDFORM地貌类型 SQKM（面积，单位平方公里）

In [19]:
import numpy as np
# 确保坐标系一致（重要！）
if not points_gdf.crs.equals(dimao.crs):
    points_gdf = points_gdf.to_crs(dimao.crs)

# 空间连接：将点与包含它的多边形关联
points_with_dimao = gpd.sjoin(
    points_gdf,
    dimao,
    how="left",
    predicate="within"  # 点必须位于多边形内部
)

In [20]:
from sklearn.preprocessing import StandardScaler
# points_gdf已包含以下字段：
# 现有字段：NDVI, 宜黄DEM, 曲率, 坡度yihuangjiangshui, NEAR_FID, yihuangjiangshui
# 新提取字段：SLOPE_MED, RELIEF_MED, ELEV_MED, DOMSOIL, LANDFORM, SQKM, PERIMETER, AREA
# 提取关联后的地貌属性（如岩性、坡度等）
selected_features = [ 'SLOPE_MED','RELIEF_MED','ELEV_MED','DOMSOIL','LANDFORM','SQKM','PERIMETER','AREA']
points_gdf = points_with_dimao[selected_features + points_gdf.columns.tolist()].copy()

In [21]:
# 1. 处理重复字段：计算局部与区域的差值
points_gdf.loc[:,'slope_diff'] = points_gdf['坡度'] - points_gdf['SLOPE_MED']
points_gdf.loc[:,'elevation_diff'] = points_gdf['宜黄dem'] - points_gdf['ELEV_MED']

In [22]:
# 2. 处理DOMSOIL（土壤类型）#土壤类型独热编码
soil_dummies = pd.get_dummies(points_gdf['DOMSOIL'], prefix='soil')
points_gdf = pd.concat([points_gdf, soil_dummies], axis=1)

In [23]:
# 3. 处理LANDFORM（地貌类型）地貌类型独热编码
landform_dummies = pd.get_dummies(points_gdf['LANDFORM'], prefix='landform')
points_gdf = pd.concat([points_gdf, landform_dummies], axis=1)

In [24]:
# 4. 形状复杂度特征
points_gdf['shape_index'] = (points_gdf['PERIMETER'] ** 2) / (4 * np.pi * points_gdf['AREA'])#由周长面积计算

In [25]:
points_gdf.columns.tolist()

['SLOPE_MED',
 'RELIEF_MED',
 'ELEV_MED',
 'DOMSOIL',
 'LANDFORM',
 'SQKM',
 'PERIMETER',
 'AREA',
 'OBJECTID',
 '正样本',
 'POINT_X',
 'POINT_Y',
 'ndvi',
 '宜黄dem',
 '曲率',
 '坡向',
 '坡度',
 'yihuangjia',
 'NEAR_FID',
 'geometry',
 'distance_to_water',
 'distance_to_road',
 'rock_IA1',
 'rock_IP4',
 'rock_SC2',
 'rock_SC4',
 'rock_SO1',
 'slope_diff',
 'elevation_diff',
 'soil_ACh',
 'soil_ACu',
 'soil_ALh',
 'soil_ATc',
 'landform_LP',
 'landform_SH',
 'landform_SM',
 'landform_TM',
 'shape_index']

In [26]:
# 5. 排除冗余字段
points_gdf.drop(columns=['NEWSUID', 'INDEX', 'ISOCSUID', 'SMU_MOD', 'PRID','PERIMETER','AREA','DOMSOIL', 'LANDFORM'], inplace=True,errors='ignore')

In [27]:
points_gdf

SLOPE_MED  RELIEF_MED  ELEV_MED         SQKM  OBJECTID  正样本     POINT_X  \
0         27.1       240.0     529.0  3339.064421         1    0  116.161667   
1         27.1       240.0     529.0  3339.064421         2    0  116.162500   
2         27.1       240.0     529.0  3339.064421         3    0  116.227222   
3         27.1       240.0     529.0  3339.064421         4    0  116.173056   
4         23.5       241.0     604.0    10.425662         5    0  116.167222   
..         ...         ...       ...          ...       ...  ...         ...   
420        4.4        32.0      87.0    25.253434       421    1  116.347500   
421        7.7        68.0     120.0    65.988743       422    1  116.350000   
422       12.5       155.0     172.0   207.668235       423    1  116.288056   
423       12.5       155.0     172.0   207.668235       424    1  116.317222   
424       12.5       155.0     172.0   207.668235       425    1  116.405556   

       POINT_Y      ndvi  宜黄dem  ...  elevation_diff  soil_ACh  soil_ACu  \
0    27.052222  0.790292    553  ...            24.0      True     False   
1    27.060556  0.791342    501  ...           -28.0      True     False   
2    27.084722  0.770125    431  ...           -98.0      True     False   
3    27.085000  0.731075    333  ...          -196.0      True     False   
4    27.098056  0.752542    439  ...          -165.0     False      True   
..         ...       ...    ...  ...             ...       ...       ...   
420  27.660556  0.630142     97  ...            10.0     False     False   
421  27.668333  0.723758    115  ...            -5.0      True     False   
422  27.631389  0.762558    133  ...           -39.0      True     False   
423  27.623889  0.788492    177  ...             5.0      True     False   
424  27.630000  0.749033    120  ...           -52.0      True     False   

     soil_ALh  soil_ATc landform_LP  landform_SH  landform_SM  landform_TM  \
0       False     False       False         True        False        False   
1       False     False       False         True        False        False   
2       False     False       False         True        False        False   
3       False     False       False         True        False        False   
4       False     False       False         True        False        False   
..        ...       ...         ...          ...          ...          ...   
420     False      True        True        False        False        False   
421     False     False        True        False        False        False   
422     False     False       False         True        False        False   
423     False     False       False         True        False        False   
424     False     False       False         True        False        False   

     shape_index  
0      57.864406  
1      57.864406  
2      57.864406  
3      57.864406  
4       2.322222  
..           ...  
420     3.338170  
421     5.487212  
422     5.460597  
423     5.460597  
424     5.460597  

[424 rows x 34 columns]

In [28]:
import pandas as pd
# 示例：将所有布尔值列转换为0和1
boolean_columns = [col for col in points_gdf.columns if points_gdf[col].dtype == 'bool']
points_gdf[boolean_columns] = points_gdf[boolean_columns].astype(int)
points_gdf

SLOPE_MED  RELIEF_MED  ELEV_MED         SQKM  OBJECTID  正样本     POINT_X  \
0         27.1       240.0     529.0  3339.064421         1    0  116.161667   
1         27.1       240.0     529.0  3339.064421         2    0  116.162500   
2         27.1       240.0     529.0  3339.064421         3    0  116.227222   
3         27.1       240.0     529.0  3339.064421         4    0  116.173056   
4         23.5       241.0     604.0    10.425662         5    0  116.167222   
..         ...         ...       ...          ...       ...  ...         ...   
420        4.4        32.0      87.0    25.253434       421    1  116.347500   
421        7.7        68.0     120.0    65.988743       422    1  116.350000   
422       12.5       155.0     172.0   207.668235       423    1  116.288056   
423       12.5       155.0     172.0   207.668235       424    1  116.317222   
424       12.5       155.0     172.0   207.668235       425    1  116.405556   

       POINT_Y      ndvi  宜黄dem  ...  elevation_diff  soil_ACh  soil_ACu  \
0    27.052222  0.790292    553  ...            24.0         1         0   
1    27.060556  0.791342    501  ...           -28.0         1         0   
2    27.084722  0.770125    431  ...           -98.0         1         0   
3    27.085000  0.731075    333  ...          -196.0         1         0   
4    27.098056  0.752542    439  ...          -165.0         0         1   
..         ...       ...    ...  ...             ...       ...       ...   
420  27.660556  0.630142     97  ...            10.0         0         0   
421  27.668333  0.723758    115  ...            -5.0         1         0   
422  27.631389  0.762558    133  ...           -39.0         1         0   
423  27.623889  0.788492    177  ...             5.0         1         0   
424  27.630000  0.749033    120  ...           -52.0         1         0   

     soil_ALh  soil_ATc landform_LP  landform_SH  landform_SM  landform_TM  \
0           0         0           0            1            0            0   
1           0         0           0            1            0            0   
2           0         0           0            1            0            0   
3           0         0           0            1            0            0   
4           0         0           0            1            0            0   
..        ...       ...         ...          ...          ...          ...   
420         0         1           1            0            0            0   
421         0         0           1            0            0            0   
422         0         0           0            1            0            0   
423         0         0           0            1            0            0   
424         0         0           0            1            0            0   

     shape_index  
0      57.864406  
1      57.864406  
2      57.864406  
3      57.864406  
4       2.322222  
..           ...  
420     3.338170  
421     5.487212  
422     5.460597  
423     5.460597  
424     5.460597  

[424 rows x 34 columns]

In [29]:
columns_to_drop = points_gdf.columns[0:34]
columns_to_drop

Index(['SLOPE_MED', 'RELIEF_MED', 'ELEV_MED', 'SQKM', 'OBJECTID', '正样本',
       'POINT_X', 'POINT_Y', 'ndvi', '宜黄dem', '曲率', '坡向', '坡度', 'yihuangjia',
       'NEAR_FID', 'geometry', 'distance_to_water', 'distance_to_road',
       'rock_IA1', 'rock_IP4', 'rock_SC2', 'rock_SC4', 'rock_SO1',
       'slope_diff', 'elevation_diff', 'soil_ACh', 'soil_ACu', 'soil_ALh',
       'soil_ATc', 'landform_LP', 'landform_SH', 'landform_SM', 'landform_TM',
       'shape_index'],
      dtype='object')

In [30]:
# 保存为CSV文件（关键参数说明）
points_gdf.to_csv(
    'processed_data.csv',          # 文件路径（可自定义）
    index=False,                   # 不保存行索引
    encoding='utf-8',              # 避免中文乱码（重要！）
    sep=','               # 列分隔符（默认逗号）
)
print("数据已保存为 processed_data.csv")

数据已保存为 processed_data.csv


In [32]:
import pandas as pd
import os

# 假设你的数据保存在 DataFrame 变量 points_gdf 中
# 1. 定义保存路径（D盘的某个文件夹）
output_path = r'D:\MachineLearningTrain\zf\processed_data.csv'  # 使用原始字符串（r''）避免转义问题

# 2. 确保目标文件夹存在，如果不存在则创建
os.makedirs(os.path.dirname(output_path), exist_ok=True)

# 3. 保存CSV文件
points_gdf.to_csv(
    output_path,          # 指定完整路径
    index=False,          # 不保存索引
    encoding='utf-8',     # 避免中文乱码
    sep=','               # 列分隔符（默认逗号）
)
print(f"文件已保存到：{output_path}")

文件已保存到：D:\MachineLearningTrain\zf\processed_data.csv
